cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,387 @@
+		"1.3.0": {
+			"1.3.1": true,
+			"1.3.2": true,
+		},
+		"1.3.1": {
+			"1.3.2": true,
+		},
+	}
+	l *logger
+)
+
+type upgradeArgs struct {
+	// istioNamespace k8s namespace for istio control plane.
+	istioNamespace string
+	// inFilename is the path to the input IstioControlPlane CR.
+	inFilename string
+	// kubeConfigPath is the path to kube config file.
+	kubeConfigPath string
+	// context is the cluster context in the kube config.
+	context string
+	// wait is flag that indicates whether to wait resources ready before exiting.
+	wait bool
+	// yes means don't ask for confirmation (asking for confirmation not implemented).
+	yes bool
+	// dryRun means running the upgrade process without actually applying the changes.
+	dryRun bool
+	// force means directly applying the upgrade without eligibility checks.
+	force bool
+}
+
+func addUpgradeFlags(cmd *cobra.Command, args *upgradeArgs) {
+	cmd.PersistentFlags().StringVarP(&args.istioNamespace, "istioNamespace",
+		"i", "istio-system", "Istio system namespace")
+	cmd.PersistentFlags().StringVarP(&args.inFilename, "filename",
+		"f", "", "Path to file containing IstioControlPlane CustomResource")
+	cmd.PersistentFlags().StringVarP(&args.kubeConfigPath, "kubeconfig",
+		"c", "", "Path to kube config")
+	cmd.PersistentFlags().StringVar(&args.context, "context", "",
+		"The name of the kubeconfig context to use")
+	cmd.PersistentFlags().BoolVarP(&args.yes, "yes", "y", false,
+		"Do not ask for confirmation")
+	cmd.PersistentFlags().BoolVarP(&args.wait, "wait", "w", false,
+		"Wait, if set will wait until all Pods, Services, and minimum number of Pods "+
+			"of a Deployment are in a ready state before the command exits. "+
+			"It will wait for a maximum duration of --readiness-timeout seconds")
+	cmd.PersistentFlags().BoolVarP(&args.dryRun, "dryrun", "d", false,
+		"Run the upgrade process without actually applying the changes")
+	cmd.PersistentFlags().BoolVar(&args.force, "force", false,
+		"Apply the upgrade without eligibility checks")
+}
+
+// Upgrade command upgrades Istio control plane in-place with eligibility checks
+func Upgrade() *cobra.Command {
+	macArgs := &upgradeArgs{}
+	cmd := &cobra.Command{
+		Use:     "upgrade",
+		Short:   "Upgrade Istio control plane in-place",
+		Example: `mesh upgrade`,
+		RunE: func(cmd *cobra.Command, args []string) (e error) {
+			return upgrade(cmd, macArgs)
+		},
+	}
+	addUpgradeFlags(cmd, macArgs)
+	return cmd
+}
+
+func upgrade(cmd *cobra.Command, args *upgradeArgs) (err error) {
+	initLog(cmd)
+	kubeClient := getKubeClient(args.kubeConfigPath, args.context)
+	currentVer := retrieveControlPlaneVersion(kubeClient, args.istioNamespace)
+	targetVer := retrieveClientVersion()
+	if !args.force {
+		checkSupportedVersions(currentVer, targetVer)
+		currentValues := readValuesFromInjectorConfigMap(kubeClient, args.istioNamespace)
+		targetValues := genValuesFromFile(targetVer, args.inFilename)
+		checkUpgradeValues(currentValues, targetValues, args.yes)
```

Comment:
agree, I will create an issue to track it for now.

Reply:
@ostromart wait, here we compares the values installed in the cluster with the values the cmd is going to apply (already having the overridden values generated). So this shouldn't output any default changes that will be overridden later.

Reply:
Yes, but this will not give the correct behavior. We want to alert user if any value will change in the cluster. That will not be the case if they have overridden (their override always has priority) but the command will report that the value may change.
What we should do is check for diffs between the base profiles, then remove anything from the resulting diffs that user has overridden.